Rust 1.45.0 release announcement (#649)
rust-lang/blog.rust-lang.org: posts/2020-07-16-Rust-1.45.0.md

not understand what that means. It's full of jargon, but there's a simpler
explanation: if you cast a floating point number that's large to an integer
that's small, you get undefined behavior.

That means that this, for example, was not well-defined:

```rust
fn cast(x: f32) -> u8 {
    x as u8
}

fn main() {
    let f = 300.0;
    
    let x = cast(f);
    
    println!("x: {}", x);
}
```

On Rust 1.44.0, this happens to print "x: 0" on my machine. But it could
print anything, or do anything: this is undefined behavior. But we have
no `unsafe` code here. This is what we call a "soundness" bug, that is,
it is a bug where the compiler does the wrong thing. We tag these bugs
as
[I-unsound](https://github.com/rust-lang/rust/issues?q=is%3Aissue+is%3Aopen+label%3A%22I-unsound+%F0%9F%92%A5%22)
on our issue tracker, and take them very seriously.

This bug took a long time to resolve, though. The reason is that it was very
unclear what the correct path forward was.

In the end, the decision was made to do this:

* `as` would perform a "saturating cast".
* A new `unsafe` cast would be added if you wanted to skip the checks.

This is very similar to array access, for example:

> * `array[i]` will check to make sure that `array` as at least `i` elements
> * You can use `unsafe { array.get_unchecked(i) }` to skip the check

Reviewer: ```suggestion
* `array[i]` will check to make sure that `array` has at least `i` elements.
* You can use `unsafe { array.get_unchecked(i) }` to skip the check.
```

So, what's a saturating cast? Let's look at a slightly modified example:

```rust
fn cast(x: f32) -> u8 {
    x as u8
}

fn main() {
    let too_big = 300.0;
    let too_small = -100.0;
    let nan = f32::NAN;
    
    println!("too_big_casted = {}", cast(too_big));
    println!("too_small_casted = {}", cast(too_small));
    println!("not_a_number_casted = {}", cast(nan));
}
```

This will print:

```text
too_big_casted = 255
too_small_casted = 0
not_a_number_casted = 0
```

That is, numbers that are too big turn into the largest possible value.
Numbers that are too small produce the smallest possible value (which is
zero). NaN produces zero.

The new API to cast in an unsafe manner is:

```rust
let x: f32 = 1.0;
let y: u8 = unsafe { x.to_int_unchecked() };
```

But as always, you should only use this method as a last resort. Just like
with array access, the compiler can often optimize the checks away, making
the safe and unsafe versions equivalent when the compiler can prove it.

> ## Stabilizing function-like procedrual macros in expressions, patterns, and statements

Reviewer 2: ```suggestion
## Stabilizing function-like procedural macros in expressions, patterns, and statements
```

In [Rust 1.30.0](https://blog.rust-lang.org/2018/10/25/Rust-1.30.0.html), we stabilized

> "function-like procedrual macros in item position." For example, [the

Reviewer 3: ```suggestion
"function-like procedural macros in item position." For example, [the
```

```rust
gobject_gen! {
    class MyClass: GObject {
        foo: Cell<i32>,
        bar: RefCell<String>,
    }

    impl MyClass {
        virtual fn my_virtual_method(&self, x: i32) {
            ... do something with x ...
        }
    }
}
```

The "in item position" bit is some jargon, but basically what this means is that
you could only invoke `gobject_gen!` in certain places in your code.

Rust 1.45.0 adds the ability to invoke procedural macros in three new places:

```rust
// imagine we have a procedural macro named "mac"

mac!(); // item position, this was what was stable before

// but these three are new:
fn main() {
  let expr = mac!(); // expression position

  match expr {
      mac!() => {} // pattern position
  }

  mac!(); // statement position
}
```

Being able to use macros in more places is interesting, but there's another
reason why many Rustaceans have been waiting for this feature for a long
time. [Rocket](https://rocket.rs/) is a popular web framework for Rust.
Initially released in December of 2016, it has often been described as one of
the best things the Rust ecosystem has to offer. Here's the "hello world"
example from its homepage:

```rust

> #![feature(proc_macro_hygiene, decl_macro)]

Reviewer 4: Would it be better to use the expected example from `0.5`?

```rust
#[macro_use] extern crate rocket;

#[get("/<name>/<age>")]
fn hello(name: String, age: u8) -> String {
    format!("Hello, {} year old named {}!", age, name)
}

#[launch]
fn rocket() -> rocket::Rocket {
    rocket::ignite().mount("/hello", routes![hello])
}
```

If so, perhaps the text above could change to:

> Being able to use macros in more places is interesting, but there's another
reason why many Rustaceans have been waiting for this feature for a long time:
[Rocket](https://rocket.rs/). Inititally released in December of 2016, Rocket is
a popular web framework for Rust often described as one of the best things the
Rust ecosystem has to offer. Here's the "hello world" example from its upcoming release:

Reviewer 5: @SergioBenitez done!

Reviewer 4: The next paragraph will also need to change to accommodate this change. Drafting a suggestion now.

#[macro_use] extern crate rocket;

#[get("/hello/<name>/<age>")]
fn hello(name: String, age: u8) -> String {
    format!("Hello, {} year old named {}!", age, name)
}

fn main() {
    rocket::ignite().mount("/", routes![hello]).launch();
}
```

> At the same time, Rocket was only able to achieve its power and ergonomics by

Reviewer 2: the "at the same time" feels weird here

Reviewer: I agree, I think the sentence works just as well without it.
```suggestion
Rocket was only able to achieve its power and ergonomics by
```

> Rocket was only able to achieve its power and ergonomics by
> using nightly-only features. You can see two of them in the example above.
> However, as you may guess from the name of those two features, these are the
> ones that are finally stable! [This
> issue](https://github.com/SergioBenitez/Rocket/issues/19) has been tracking
> the history of all of the various features, and now they're all checked off!
> 
> Rocket 0.5 has not yet been released, but when it is, many folks will be very
> happy :)

Reviewer 4: ```suggestion
Until today, Rocket depended on nightly-only features to deliver on its promise
of flexibility and ergonomics. In fact, as can be seen on the [project's
homepage](https://rocket.rs/v0.4), the same example above in the current version
of Rocket requires the `proc_macro_hygiene` feature to compile. However, as you
may guess from the feature's name, today it ships in stable! [This
issue](https://github.com/SergioBenitez/Rocket/issues/19) tracked the history of
nightly-only features in Rocket. Now, they're all checked off!

This next version of Rocket is still in the works, but when released, many folks
will be very happy :)
```
